### PR TITLE
NF: AFNI 3dNetCorr as afni.NetCorr

### PR DESCRIPTION
**Pull request by aghayoor**
This is a WIP to afni NetCorr to nipype

### CODE REVIEW

**Comment by aghayoor** [comment]
@effigies : Hi Chris, this is WIP commit to add afni NetCorr to nipype.

Everything works well when I call the interface like this:

```
ncorr = afni.NetCorr()
ncorr.inputs.in_file = '<workflowDir>/bandpass/bold_trimmed_wmcsfreg_bp.nii'
ncorr.inputs.mask = mniBrainMask
ncorr.inputs.in_rois = mniLabels
ncorr.inputs.ts_wb_corr = True
ncorr.inputs.ts_wb_Z = True
ncorr.inputs.fish_z = True
ncorr.inputs.nifti = True
ncorr.inputs.out_file = '<workflowDir>/correlation/sub0.tp1.ncorr'
ncorr.run()
```

However, when I wrap this interface in a MapNode, the workflow runs with no error, but it does not write the outputs.

```
ncorr = pe.Node(interface=afni.NetCorr(), name='correlation')
ncorr.inputs.ts_out = True
ncorr.inputs.ts_label = True
ncorr.inputs.ts_wb_strlabel = True
ncorr.inputs.nifti = True
ncorr.inputs.ts_wb_corr = True
ncorr.inputs.ts_wb_Z = True
ncorr.inputs.fish_z = True
ncorr.inputs.out_file = 'sub0.tp1.ncorr'
workflow.connect(bandpass, 'out_file', ncorr, 'in_file')
workflow.connect(inputsSpec, 'inputMNILables', ncorr, 'in_rois')
workflow.connect(inputsSpec, 'inputMNIBrainMask', ncorr, 'mask')
```

Do you have any advice on this?

**Comment by effigies** [comment]
The outputs should be in the workflow's working directory... Can you show the `tree` of that directory?

**Comment by aghayoor** [comment]
> The outputs should be in the workflow's working directory... Can you show the `tree` of that directory?

Here is the tree:
Screen Shot 2020-10-26 at 2 51 32 PM

I expected to see the output .netcc file and correlation maps in the correlation directory.

**Comment by effigies** [comment]
This may be a case where unused outputs are not saved. Try using the configuration option `remove_unnecessary_outputs`. See https://miykael.github.io/nipype_tutorial/notebooks/basic_execution_configuration.html.

**Comment by aghayoor** [comment]
> This may be a case where unused outputs are not saved. Try using the configuration option `remove_unnecessary_outputs`. See https://miykael.github.io/nipype_tutorial/notebooks/basic_execution_configuration.html.

Well that was the case here. Thank you!
Is it possible to avoid deleting those files without changing the configuration?

**Comment by effigies** [comment]
If the output is used as the input to another node, then it won't be deleted. While you're debugging a workflow, it's useful to have on, and then go back to the default when you're in production.

**Comment by aghayoor** [comment]
@effigies : Thanks! Can you please review and merge this pull request?

File: nipype/interfaces/afni/preprocess.py [hunk 1]
```diff
@@ -2556,6 +2556,79 @@ def _format_arg(self, name, trait_spec, value):
...
+    output_mask_nonnull = traits.Bool(argstr="-output_mask_nonnull")
+    push_thru_many_zeros = traits.Bool(argstr="-push_thru_many_zeros")
+    ignore_LT = traits.Bool(argstr="-ignore_LT")
+    out_file = File(
+        name_template="%s_netcorr",
+        desc="output file name part",
+        argstr="-prefix %s",
+        position=1,
+        name_source="in_file",
+    )
```
Comment (effigies) [inline nipype/interfaces/afni/preprocess.py:2582]: These should all have descriptions explaining what they do. And feel free to use more descriptive names than the 3dNetCorr flags.

File: nipype/interfaces/afni/preprocess.py [hunk 2]
```diff
@@ -2556,6 +2556,79 @@ def _format_arg(self, name, trait_spec, value):
...
+
+class NetCorrOutputSpec(TraitedSpec):
+    out_matrix = File(desc="output text file for correlation stats")
+
+class NetCorr(AFNICommand):
+    """Calculate correlation matrix of a set of ROIs (using mean time series of
+    each). Several networks may be analyzed simultaneously, one per brick.
+
+    For complete details, see the `3dTcorrMap Documentation.
+    <https://afni.nimh.nih.gov/pub/dist/doc/program_help/3dNetCorr.html>`_
```
Comment (effigies) [inline nipype/interfaces/afni/preprocess.py:2592]:
```suggestion
    For complete details, see the `3dNetCorr Documentation
    <https://afni.nimh.nih.gov/pub/dist/doc/program_help/3dNetCorr.html>`_.
```

File: nipype/interfaces/afni/preprocess.py [hunk 3]
```diff
@@ -2556,6 +2556,79 @@ def _format_arg(self, name, trait_spec, value):
...
+
+    Examples
+    --------
+    >>> from nipype.interfaces import afni
+    >>> ncorr = afni.NetCorr()
+    >>> ncorr.inputs.in_file = 'functional.nii'
+    >>> ncorr.inputs.mask = 'mask.nii'
+    >>> ncorr.inputs.in_rois = 'rois.nii'
```
Comment (effigies) [inline nipype/interfaces/afni/preprocess.py:2600]: This file needs to exist in [nipype.testing/data](https://github.com/nipy/nipype/tree/master/nipype/testing/data).

File: nipype/interfaces/afni/preprocess.py [hunk 4]
```diff
@@ -2556,6 +2556,79 @@ def _format_arg(self, name, trait_spec, value):
...
+    >>> ncorr.inputs.ts_wb_corr = True
+    >>> ncorr.inputs.ts_wb_Z = True
+    >>> ncorr.inputs.fish_z = True
+    >>> ncorr.inputs.prefix = 'sub0.tp1.ncorr'
```
Comment (effigies) [inline nipype/interfaces/afni/preprocess.py:2604]:
```suggestion
    >>> ncorr.inputs.out_file = 'sub0.tp1.ncorr'
```

File: nipype/interfaces/afni/preprocess.py [hunk 5]
```diff
@@ -2556,6 +2556,79 @@ def _format_arg(self, name, trait_spec, value):
...
+    >>> ncorr.cmdline # doctest: +SKIP
```
Comment (effigies) [inline nipype/interfaces/afni/preprocess.py:2605]: It's worth testing. You may need to fix up the order, but it should be stable once you've set it.

```suggestion
    >>> ncorr.cmdline
```

File: nipype/interfaces/afni/preprocess.py [hunk 6]
```diff
@@ -2556,6 +2556,79 @@ def _format_arg(self, name, trait_spec, value):
...
+            prefix = self._gen_fname(self.inputs.in_file, suffix="_netcorr")
+        else:
+            prefix = self.inputs.out_file
+
+        # All outputs should be in the same directory as the prefix
+        out_dir = os.path.dirname(os.path.abspath(prefix))
+
+        outputs["out_matrix"] = (
+            fname_presuffix(prefix, suffix="_000", use_ext=False, newpath=out_dir) + ".netcc"
+        )
```
Comment (effigies) [inline nipype/interfaces/afni/preprocess.py:2628]: What is the output file called? I would expect, if I set `netcorr.inputs.out_file`, the output file is called exactly that.

**Comment by effigies** [comment]
@aghayoor Just a reminder that I left a review when you have time.

**Comment by effigies** [comment]
Hi @aghayoor, any chance of finishing this one up before a release?

**Comment by aghayoor** [comment]
Hi @effigies , I was out of office for about two weeks. Thanks for your comments. I will try to address all of your comments by the end of this week.

File: nipype/interfaces/afni/preprocess.py (hunk 6 above)
Comment (aghayoor) [reply nipype/interfaces/afni/preprocess.py:2731]: @effigies : the `netcorr.inputs.out_file` defines the prefix for the output.
Afni 3dNetCorr doesn't have actual output options, but it uses the prefix to write out the correlation matrix and correlation maps to the output directory.
This added function finds and returns the convention filename for the output correlation matrix.

I am also interested to return the correlation maps return to a subdirectory in the output directory. The number of these correlation maps depends on the number of regions in the input ROI label map. Do you have any idea how I can those to the output as well?

**Comment by aghayoor** [comment]
@effigies Hi, I addressed your comments. Could please review this new PR?

File: nipype/interfaces/afni/preprocess.py [hunk 7]
```diff
@@ -2556,6 +2556,182 @@ def _format_arg(self, name, trait_spec, value):
...
+             "user will really, really, really need to address their masking",
+        argstr="-push_thru_many_zeros")
+    ignore_LT = traits.Bool(
+        desc="switch to ignore any label table labels in the "
+             "'-in_rois' file, if there are any labels attached",
+        argstr="-ignore_LT")
+    out_file = File(
+        desc="output file name part",
+        name_template="%s_netcorr",
+        desc="output file name part",
```
Comment (effigies) [inline nipype/interfaces/afni/preprocess.py:2681]:
```suggestion
```

**Comment by aghayoor** [comment]
@effigies : addressed your comment. Thanks for catching that.

File: nipype/interfaces/afni/preprocess.py [hunk 8]
```diff
@@ -2556,6 +2556,181 @@ def _format_arg(self, name, trait_spec, value):
...
+    >>> ncorr = afni.NetCorr()
+    >>> ncorr.inputs.in_file = 'functional.nii'
+    >>> ncorr.inputs.mask = 'mask.nii'
+    >>> ncorr.inputs.in_rois = 'maps.nii'
+    >>> ncorr.inputs.ts_wb_corr = True
+    >>> ncorr.inputs.ts_wb_Z = True
+    >>> ncorr.inputs.fish_z = True
+    >>> ncorr.inputs.out_file = 'sub0.tp1.ncorr'
+    >>> ncorr.cmdline
+    '3dNetCorr -prefix sub0.tp1.ncorr -inset functional.nii -mask mask.nii -in_rois maps.nii -ts_wb_corr -ts_wb_Z -fish_z'
```
Comment (effigies) [inline nipype/interfaces/afni/preprocess.py:2708]: For the things without explicit positions, the arguments get sorted, IIRC. In any event, I think you want:

```suggestion
    '3dNetCorr -prefix sub0.tp1.ncorr -fish_z -inset functional.nii -in_rois maps.nii -mask mask.nii -ts_wb_Z -ts_wb_corr'
```

Comment (aghayoor) [reply nipype/interfaces/afni/preprocess.py]: @effigies: Thanks. I applied your suggestion. How the arguments are sorted?

Comment (effigies) [reply nipype/interfaces/afni/preprocess.py]: https://github.com/nipy/nipype/blob/47fe00b387e205f3f884691a208ef7f20df4119f/nipype/interfaces/base/core.py#L967-L1008

And the heart is:

https://github.com/nipy/nipype/blob/47fe00b387e205f3f884691a208ef7f20df4119f/nipype/interfaces/base/core.py#L983

**Review by effigies — APPROVED** [review]

File: nipype/interfaces/afni/preprocess.py (hunk 6 above)
Comment (aghayoor) [reply nipype/interfaces/afni/preprocess.py:2730]: @effigies : Thanks for the approval.
also, do you have any idea on this?
Please let me know if my question needs more clarification

Comment (effigies) [reply nipype/interfaces/afni/preprocess.py:2730]: Ah, sorry, missed this. Assuming these files exist, you can use `glob` to get them. See an example here:

https://github.com/nipy/nipype/blob/e9217c299149be829a985d8f822c3a91e29b2f4b/nipype/interfaces/fsl/model.py#L459-L490

Also, can you run `make specs` and commit the new file?

Comment (aghayoor) [reply nipype/interfaces/afni/preprocess.py:2730]: Thanks! That was very helpful.

**Comment by aghayoor** [comment]
Hi @effigies , I added correlation maps to the output of this module and I also committed the output file of `make specs`. Please review and let me know if you have any comments.

**Comment by aghayoor** [comment]
Hi @effigies , just a reminder that I updated the PR. Please review when you have time. Thanks!

**Comment by effigies** [comment]
Thanks for the bump!

**Pull request merged**